src/rmp/src/logic_cut.cpp

```
@@ -0,0 +1,292 @@
+  abc::Mio_Gate_t* current_gate = abc::Mio_LibraryReadGates(library);
+  while (current_gate) {
+    result[abc::Mio_GateReadName(current_gate)] = current_gate;
+    current_gate = abc::Mio_GateReadNext(current_gate);
+  }
+
+  return result;
+}
+
+std::unordered_map<abc::Mio_Gate_t*, std::vector<std::string>>
+MioGateToPortOrder(abc::Mio_Library_t* library)
+{
+  std::unordered_map<abc::Mio_Gate_t*, std::vector<std::string>> result;
+
+  // Create map of cell names to Mio_Gate_t
+  abc::Mio_Gate_t* current_gate = abc::Mio_LibraryReadGates(library);
+  while (current_gate) {
+    abc::Mio_Pin_t* current_pin = abc::Mio_GateReadPins(current_gate);
+    std::vector<std::string> pin_order;
+    while (current_pin) {
+      pin_order.emplace_back(abc::Mio_PinReadName(current_pin));
+      current_pin = abc::Mio_PinReadNext(current_pin);
+    }
+    result[current_gate] = std::move(pin_order);
+    current_gate = abc::Mio_GateReadNext(current_gate);
+  }
+
+  return result;
+}
+
+std::unordered_map<sta::Instance*, abc::Abc_Obj_t*> CreateStandardCells(
+    const std::unordered_set<sta::Instance*>& cut_instances,
+    abc::Abc_Ntk_t& abc_network,
+    sta::dbNetwork* network,
+    abc::Mio_Library_t* library,
+    utl::Logger* logger)
+{
+  std::unordered_map<std::string, abc::Mio_Gate_t*> cell_name_to_mio
+      = NameToAbcGateMap(library);
+
+  std::unordered_map<sta::Instance*, abc::Abc_Obj_t*> instance_map;
+  for (sta::Instance* instance : cut_instances) {
+    abc::Abc_Obj_t* abc_cell = abc::Abc_NtkCreateNode(&abc_network);
+
+    // Assign this node its standard cell. This is what makes this node
+    // an AND gate or whatever.
+    sta::LibertyCell* cell = network->libertyCell(instance);
+    std::string cell_name = cell->name();
+    if (cell_name_to_mio.find(cell_name) == cell_name_to_mio.end()) {
+      logger->error(utl::RMP,
+                    1001,
+                    "cell: {} was not found in the ABC library, this is a bug.",
```

", this is a bug" isn't useful.

---

I think I'm trying to communicate that this should never happen and should be reported. i.e. there's nothing a user can do, it's a bug. How do you think we should best communicate that?

---

"Please report this internal error."

---

Done.